packages/react-dom/src/__tests__/ReactDOMHydrationDiff-test.js

```diff
@@ -0,0 +1,639 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+'use strict';
+
+let React;
+let ReactDOMClient;
+let ReactDOMServer;
+let act;
+
+const util = require('util');
+const realConsoleError = console.error;
+
+describe('ReactDOMServerHydration', () => {
+  let container;
+
+  beforeEach(() => {
+    jest.resetModules();
+    React = require('react');
+    ReactDOMClient = require('react-dom/client');
+    ReactDOMServer = require('react-dom/server');
+    act = require('react-dom/test-utils').act;
+
+    console.error = jest.fn();
+    container = document.createElement('div');
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+    console.error = realConsoleError;
+  });
+
+  function normalizeCodeLocInfo(str) {
+    return (
+      typeof str === 'string' &&
+      str.replace(/\n +(?:at|in) ([\S]+)[^\n]*/g, function(m, name) {
+        return '\n    in ' + name + ' (at **)';
+      })
+    );
+  }
+
+  function formatMessage(args) {
+    const [format, ...rest] = args;
+    if (format instanceof Error) {
+      return 'Caught [' + format.message + ']';
+    }
+    if (format.indexOf('Error: Uncaught [') === 0) {
+      // Ignore errors captured by jsdom and their stacks.
+      // We only want console errors in this suite.
+      return null;
+    }
+    rest[rest.length - 1] = normalizeCodeLocInfo(rest[rest.length - 1]);
+    return util.format(format, ...rest);
+  }
+
+  function formatConsoleErrors() {
+    return console.error.mock.calls.map(formatMessage).filter(Boolean);
+  }
+
+  function testMismatch(Mismatch) {
+    const htmlString = ReactDOMServer.renderToString(
+      <Mismatch isClient={false} />,
+    );
+    container.innerHTML = htmlString;
+    act(() => {
+      ReactDOMClient.hydrateRoot(container, <Mismatch isClient={true} />);
+    });
+    return formatConsoleErrors();
+  }
+
+  describe('text mismatch', () => {
+    // @gate __DEV__
+    // @gate enableClientRenderFallbackOnHydrationMismatch
```

Comment:
I'm gating all of these on the new flag because the messages are different with it. But maybe it's worth snapshotting both versions to make sure I don't introduce regressions?